lib/vm/src/trap/traphandlers.rs

```diff
@@ -753,13 +818,17 @@ impl<T> TrapHandlerContextInner<T> {
             return false;
         }
 
-        let signal_trap = maybe_fault_address.map(|addr| {
-            if self.coro_trap_handler.stack_ptr_in_bounds(addr) {
-                TrapCode::StackOverflow
-            } else {
-                TrapCode::HeapAccessOutOfBounds
-            }
-        });
+        let signal_trap = if trap_code.is_some() {
+            trap_code
+        } else {
+            maybe_fault_address.map(|addr| {
+                if self.coro_trap_handler.stack_ptr_in_bounds(addr) {
+                    TrapCode::StackOverflow
+                } else {
+                    TrapCode::HeapAccessOutOfBounds
+                }
+            })
+        };
```

Comment:
This will simplify code a bit:

```suggestion
        let signal_trap = trap_code.unwrap_or_else(|| {
            maybe_fault_address.map(|addr| {
                if self.coro_trap_handler.stack_ptr_in_bounds(addr) {
                    TrapCode::StackOverflow
                } else {
                    TrapCode::HeapAccessOutOfBounds
                }
            })
        });
```

Reply:
Oh, sure, i'll do that

Reply:
Ah, in fact no, I don't want to unwrap yet.

Reply:
> Ah, in fact no, I don't want to unwrap yet.

Not sure if I understood. I think the suggestion is syntactically the same, just easier to read? Or is there anything I missed?

Reply:
`signal_trap`, like `trap_code` is an `Option<TrapCode>`, not a `TrapCode` as `unwrap_...` will produce.

Reply:
Oh, got it... then I think the operation should be `or_else`. Let me know if that works.

```suggestion
        let signal_trap = trap_code.or_else(|| {
            maybe_fault_address.map(|addr| {
                if self.coro_trap_handler.stack_ptr_in_bounds(addr) {
                    TrapCode::StackOverflow
                } else {
                    TrapCode::HeapAccessOutOfBounds
                }
            })
        });
```

Reply:
Ah, yes, that works :)